### PR TITLE
System Identification Mode

### PR DESCRIPTION
**Pull request by lthall**
This is a new mode that generates chirp waveforms and injects them into various parts of the aircrafts controllers. This may then be used to generate computational models of the aircraft. These models can then be used to do offline tuning.

### CODE REVIEW

File: ArduCopter/Parameters.cpp [hunk 1]
```diff
@@ -1009,6 +1015,7 @@ ParametersG2::ParametersG2(void)
 #if AUTOTUNE_ENABLED == ENABLED
     ,autotune_ptr(&copter.autotune)
 #endif
+    ,mode_systemid_ptr(&copter.mode_systemid)
```
Comment (khancyr) [inline ArduCopter/Parameters.cpp]: miss **if MODE_SYSTEMID_ENABLED == ENABLED**

**Comment by bnsgeyer** [comment]
@lthall Great work here!  A couple of things:

1) It looks like this does not include heli.  Your sysid inputs in the attitude control library are only in the multi.cpp file.  certainly heli's can get into the mode but the inputs made into the attitude control library won't get there for heli.

2) you put the disturbance rejection signal on the angular rate just before the rate PID's.  this seems to make sense to look at how good the PID is resisting rate disturbances.  But in the courses I've taken on sysid, they have the disturbance rejection input made on the attitude feedback loop and not the rate feedback loop.  I know you had worked with some of the U.S. Army Engineers and was wondering if this implementation stemmed from that work.  I can check with some of my contacts.  Maybe we just add another set of inputs for attitude disturbance rejection?

3) My last concern is with logging.  You log the Attitude which logs the ATT, RATE, and PID messages.  the one signal I think you are missing is the pilot input signals.  I will have to think about the disturbance rejection.  you log the target vel in the PID message. So I think you are covered there.   the other concern with logging is the sample rate.  I know my controllers can't keep up with a 400 hz sample rate.  Not sure if that is the board or the SD card.  Regardless, 400 hz is not needed for this analysis, at least not for heli's.  Can we make this adjustable?  I know how I would do it,  especially since you are calling the logging from the mode, and it could be a user settable parameter.

File: ArduCopter/Parameters.cpp (hunk 1 above)
Comment (lthall) [reply ArduCopter/Parameters.cpp]: Thanks, done.

**Comment by lthall** [comment]
Hi @bnsgeyer,
1. Yeh, I didn't want to mess with Heli without your input.
2. Yes, this is a conversation we had and this approach is not sufficient for the multi feedback controller we use. We would need to provide a matching rate and attitude input that matched both each other and the correction made by the aircraft. So we can't do this using the feedback paths, however we can do this using the disturbance rejection input paths (FF turned off).
Yes, it was done this way after detailed conversations with a client with extensive experience in this area.

3. Yeh, it is just your SD card but I agree there is scope to reduce the logging rate to reduce the data collected. However, while we don't require this bandwidth to record the response, by using the full rate logging we remove the aliased noise from folding down into our data.

**Comment by lthall** [comment]
Ok, I think this is all ready to go in now. I have not reduced the logging rate in the interests of minimising parameters. I will look at adding a sub sampling factor based on the fast, medium log rate settings in the future. I will not do this now in the interests of making it available for the next release.

**Comment by bnsgeyer** [comment]
@lthall and @rmackay9   If you can give me a few days, I would like to get the changes in to get this incorporated into tradheli.

**Comment by bnsgeyer** [comment]
@lthall  is there anything that would prevent me from running this in SITL with heli.  I've added the necessary statements to the attitude control library.  I'm getting zeros on all of the parameters when SITL comes up.  then if I try to switch to the sysID mode, it crashes SITL.

**Comment by lthall** [comment]
No, I don't think so.

File: ArduCopter/mode_systemid.cpp [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+ * Init and run calls for systemId, flight mode
+ */
+
+const AP_Param::GroupInfo ModeSystemId::var_info[] = {
+
+    // @Param: _AXIS
+    // @DisplayName: System identification axis
+    // @Description: Controls which axis are being excited
+    // @User: Standard
+    // @Values: 0:None, 1:Input Roll Angle, 2:Input Pitch Angle, 3:Input Yaw Angle, 4:Recovery Roll Angle, 5:Recovery Pitch Angle, 6:Recovery Yaw Angle, 7:Disturbance Roll, 8:Disturbance Pitch, 9:Disturbance Yaw, 10:Rate Roll, 11:Rate Pitch, 12:Rate Yaw, 13:Mixer Roll, 14:Mixer Pitch, 15:Mixer Yaw, 16:Mixer Thrust
```
Comment (bnsgeyer) [inline ArduCopter/mode_systemid.cpp]: @lthall you indicate here that the user has 16 available input types, but the enumeration above only covers 13?

**Comment by bnsgeyer** [comment]
@lthall here is a link to a commit that will fix the parameters showing zeros instead of defaults 
[parameter default fix](https://github.com/bnsgeyer/ardupilot/commit/6cb1003e7cdd22205aed19fb93a81877254996db)

It is still crashing the controller when i switch to mode 25.  there is an error that shows in the arducopter window.

File: ArduCopter/mode_systemid.cpp [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+
+    AP::logger().Write(
+        "SID",
+        "TimeUS,Ax,FSt,FSp,TFin,TC,TR,TFout",
+        "s--ssssss",
+        "F--------",
+        "Qdfffffff",
+        AP_HAL::micros64(),
+        (uint8_t)systemID_axis,
+        (double)magnitude,
```
Comment (bnsgeyer) [inline ArduCopter/mode_systemid.cpp]: why are you casting these to double?  why is that necessary?

Comment (lthall) [reply ArduCopter/mode_systemid.cpp]: To be honest I am not sure. This seems to be the way it is done even when they are floats.

File: ArduCopter/mode_systemid.cpp [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+    case AP_Motors::SpoolState::SPOOLING_UP:
+    case AP_Motors::SpoolState::SPOOLING_DOWN:
+        // do nothing
+        break;
+    }
+
+    // get pilot's desired throttle
+    float pilot_throttle_scaled = get_pilot_desired_throttle();
+
+    if((fStart > 0.0f) && (fStop > 0.0f))
```
Comment (bnsgeyer) [inline ArduCopter/mode_systemid.cpp]: @lthall  or @rmackay9 What is the if statement operating on?  Just the line after it or both lines?  Leonard, what was your intention with this if statement.

**Comment by bnsgeyer** [comment]
@lthall and @rmackay9  Do NOT merge this.  I thought it was good.  ran some tests in the sitl with the build in dynamic models and it seemed to all work ok.  I got home and ran it with Real flight and the controller keeps crashing in the middle or at the end of the sysID.

File: ArduCopter/mode_systemid.cpp (hunk 4 above)
Comment (lthall) [reply ArduCopter/mode_systemid.cpp]: Ahh crap

**Comment by lthall** [comment]
I think I have found the issues and I have also added subsampled logging.

File: ArduCopter/mode_systemid.cpp [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+            (double)(delta_velocity.x / delta_velocity_dt),
+            (double)(delta_velocity.y / delta_velocity_dt),
+            (double)(delta_velocity.z / delta_velocity_dt));
+    }
+    // Full rate logging of attitude, rate and pid loops
+    copter.Log_Write_Attitude();
+}
+
+// init_test - initialises the test
+float ModeSystemId::waveform(float time)
```
Comment (rmackay9) [inline ArduCopter/mode_systemid.cpp]: it would be nice to have more comments about what this function does.  I don't mind the name but maybe consider calling it "get_waveform"?

File: ArduCopter/mode_systemid.cpp [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+        window = 0.5 - 0.5 * cosf(M_PI * time / tFadeIn);
+    } else if (time <= tRec - tFadeOut) {
+        window = 1.0;
+    } else if (time <= tRec) {
+        window = 0.5 - 0.5 * cosf(M_PI * (time - (tRec - tFadeOut)) / tFadeOut + M_PI);
+    } else {
+        window = 0.0;
+    }
+
+    if(time <= 0.0f) {
```
Comment (rmackay9) [inline ArduCopter/mode_systemid.cpp]: let's add a single space after all these "if"s.

File: ArduCopter/mode_systemid.cpp [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+            log_subsample = 2;
+        } else if(copter.should_log(MASK_LOG_ATTITUDE_MED)){
+            log_subsample = 4;
+        } else {
+            log_subsample = 8;
+        }
+    }
+    log_subsample -= 1;
+
+    log_data();
```
Comment (rmackay9) [inline ArduCopter/mode_systemid.cpp]: if we're going to log at the full loop rate I think it would be good to do the logging the "hard way".  I'm pretty sure that @WickedShell has found that the "easy way" consumes significant CPU which matters if it's going to be run at 400+hz.

File: ArduCopter/mode_systemid.cpp [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+}
+
+// init_test - initialises the test
+void ModeSystemId::log_data()
+{
+    int8_t index = copter.EKF2.getPrimaryCoreIMUIndex();
```
Comment (rmackay9) [inline ArduCopter/mode_systemid.cpp]: we probably shouldn't hard-code in this dependency on the EKF2 because some (like Solo users) regularly use the EKF3.  Instead use the AP_AHRS_NavEKF::get_primary_IMU_index() function which should return the same thing.

File: ArduCopter/mode_systemid.cpp [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+            (double)waveformFreqRads / (2 * M_PI),
+            (double)(degrees(delta_angle.x / delta_angle_dt)),
+            (double)(degrees(delta_angle.y / delta_angle_dt)),
+            (double)(degrees(delta_angle.z / delta_angle_dt)),
+            (double)(delta_velocity.x / delta_velocity_dt),
+            (double)(delta_velocity.y / delta_velocity_dt),
+            (double)(delta_velocity.z / delta_velocity_dt));
+    }
+    // Full rate logging of attitude, rate and pid loops
+    copter.Log_Write_Attitude();
```
Comment (rmackay9) [inline ArduCopter/mode_systemid.cpp]: The attitude logging can already run at 400hz.  Is there any reason why we're not using the fourhundred_hz_logging() function to do the logging?  The advantage of using the existing fourhundred_hz_logging() method is that we don't have to try as hard to ensure we don't do it twice and also it means the logging is done outside the fast loop.  It may miss some logs but the main loop will run more regularly.

File: ArduCopter/mode_systemid.cpp [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+
+    if(time <= 0.0f) {
+        waveformFreqRads = wMin;
+        output = 0.0f;
+    } else if (time <= tConst) {
+        waveformFreqRads = wMin;
+        output = window * magnitude_scale * magnitude * sinf(wMin * time - wMin * tConst);
+    } else if (time <= tRec) {
+        waveformFreqRads = wMin * expf(B * (time - tConst) / (tRec - tConst));
+        output = window * magnitude_scale * magnitude * sinf((wMin * (tRec - tConst) / B) * (expf(B * (time - tConst) / (tRec - tConst)) - 1));
```
Comment (rmackay9) [inline ArduCopter/mode_systemid.cpp]: Do we need to check that tRec and tConst are not the same which (if they were) could result in a divide by zero?

File: ArduCopter/defines.h [hunk 11]
```diff
@@ -94,7 +94,8 @@ enum tuning_func {
     TUNING_RATE_YAW_FF =                54, // body frame yaw rate controller FF term
     TUNING_RATE_MOT_YAW_HEADROOM =      55, // motors yaw headroom minimum
     TUNING_RATE_YAW_FILT =              56, // yaw rate input filter
-    TUNING_WINCH =                      57  // winch control (not actually a value to be tuned)
+    TUNING_WINCH =                      57, // winch control (not actually a value to be tuned)
+    SYSTEM_ID_MAG =                     58  // magnitude of the system ID signal
```
Comment (rmackay9) [inline ArduCopter/defines.h]: if we could rename this "_MAG" to something else that might be nice. Like "MAGNITUDE" or something like that.

File: ArduCopter/mode.h [hunk 12]
```diff
@@ -1171,6 +1173,73 @@ class ModeStabilize_Heli : public ModeStabilize {
...
 
+class ModeSystemId : public Mode {
+
+public:
+        ModeSystemId(void);
+
+        bool init(bool ignore_checks) override;
+        void run() override;
+
+        bool requires_GPS() const override { return true; }
```
[the remark this answers is not in the document]
Comment (lthall) [reply ArduCopter/mode.h]: Fixed the indenting.

File: ArduCopter/mode.h [hunk 13]
```diff
@@ -1171,6 +1173,73 @@ class ModeStabilize_Heli : public ModeStabilize {
...
+        bool has_manual_throttle() const override { return false; }
+        bool allows_arming(bool from_gcs) const override { return true; };
+        bool is_autopilot() const override { return false; }
+        bool stop_attitude_logging() const override { return true; }
+
+    static const struct AP_Param::GroupInfo var_info[];
+    void        setMagnitudeScale(float input) {magnitude_scale = input;}
```
[the remark this answers is not in the document]
Comment (lthall) [reply ArduCopter/mode.h]: I have removed the magnitude_scale and directly assigned the magnitude.

I have changed the function call to set_magnitude.

File: ArduCopter/mode.h [hunk 14]
```diff
@@ -1171,6 +1173,73 @@ class ModeStabilize_Heli : public ModeStabilize {
...
+    AP_Float    tRec;           // Time taken to complete the chirp waveform
+    AP_Float    tFadeOut;       // Time to reach zero amplitude after chirp finishes
+
+    bool        orig_bf_feedforward;
+    float       waveformTime = 0.0f;
+    float       waveformSample = 0.0f;
+    float       magnitude_scale = 1.0f;
+    float       waveformFreqRads;       // Frequency at the start of the chirp
+    float       tConst;                 // Time at constant frequency before chirp starts
+    int8_t     log_subsample;          // Subsample multiple for logging.
```
[the remark this answers is not in the document]
Comment (lthall) [reply ArduCopter/mode.h]: Fixed

File: ArduCopter/mode.h [hunk 15]
```diff
@@ -1171,6 +1173,73 @@ class ModeStabilize_Heli : public ModeStabilize {
...
+protected:
+
+    const char *name() const override { return "SYSTEMID"; }
+    const char *name4() const override { return "SYSI"; }
+
+private:
+    void        log_data();
+    float       waveform(float time);
+
+    enum AxisType {
```
[the remark this answers is not in the document]
Comment (lthall) [reply ArduCopter/mode.h]: When I changed this to enum class I got additional errors so I will need to look into this further.

File: ArduCopter/mode_systemid.cpp (hunk 8 above)
Comment (lthall) [reply ArduCopter/mode_systemid.cpp]: I think that is private but I have used the gyro and accel version. So I think I got this right.

File: ArduCopter/mode_systemid.cpp (hunk 10 above)
Comment (lthall) [reply ArduCopter/mode_systemid.cpp]: Nice catch. Done.

File: ArduCopter/defines.h (hunk 11 above)
Comment (lthall) [reply ArduCopter/defines.h]: Done.

**Comment by lthall** [comment]
Ok, I think I need some help with the logging issues.

**Comment by bnsgeyer** [comment]
@lthall I think I could help with that but not sure I could get it done before the weekend.  trying to get the heli harmonic notch stuff working but having some issue.  Was going to do another flight test tonight but if that doesn't work I may have Randy merge it and I will figure it out later.  

So if you want me to look at this.  I could probably do it and may have it completed by saturday.  Not sure if that meets the timeline?

**Comment by lthall** [comment]
@bnsgeyer you focus on the notch mate. I would love to see that go in fully supporting Heli! I am sure Randy will give me a hand with the logging stuff. I probably just need some pointers from him.

**Comment by lthall** [comment]
Ok, I think I got most of the logging change done but I am getting an error on sitl. I am also unable to make the "enum class" change work.

But if I can get some help with this I think we are pretty much there.

**Comment by lthall** [comment]
Ok, with Andy's help I have the logging going.

File: ArduCopter/mode.h (hunk 15 above)
[the remark this answers is not in the document]
Comment (peterbarker) [reply ArduCopter/mode.h]: Having NONE as a symbol in your namespace is somewhat contra-indicated.

What errors did you get?

Each instance of the use of a value from the enumeration should have changed to (e.g.)
`AxisType::INPUT_ROLL` within the class and `ModeSystemId::AxisType::INPUT_ROLL` outside the class (not possible in this case due to being private, I guess)

Definitely not blocking here.

**Comment by rmackay9** [comment]
I've squashed the commits and added two commits in this branch: https://github.com/rmackay9/rmackay9-ardupilot/commits/lthall-systemid-mode

The first commit should be pretty uncontroversial as it's mostly formatting and naming changes.  The 2nd commit changes how the logging is done slightly.

I think with these changes it looks ok to me.

**Comment by rmackay9** [comment]
Merged, thanks!
